dartagnan/src/main/java/com/dat3m/dartagnan/program/analysis/alias/InclusionBasedPointerAnalysis.java

```diff
@@ -0,0 +1,977 @@
+        final int right = reduceAbsGCD(r.alignment);
+        if (left == 0 && right == 0) {
+            return offset == 0;
+        }
+        final int divisor = left == 0 ? right : right == 0 ? left : IntMath.gcd(left, right);
+        final boolean nonNegativeIndexes = left == 0 ? offset <= 0 : right != 0 || offset >= 0;
+        return nonNegativeIndexes && offset % divisor == 0;
+    }
+
+    // Tries to insert an element into a set of edges.
+    // If the edge is already covered by the set, this operation has no effect and returns false.
+    // If the edge is inserted, any existing elements that are covered by it are removed to form a set of minimal elements.
+    private boolean addInto(List<IncludeEdge> set, IncludeEdge element) {
+        //NOTE The Stream API is too costly here
+        for (final IncludeEdge o : set) {
+            if (element.source.equals(o.source) && includes(o.modifier, element.modifier)) {
+                failedAddInto++;
+                return false;
+            }
+        }
+        set.removeIf(o -> element.source.equals(o.source) && includes(element.modifier, o.modifier));
+        set.add(element);
+        succeededAddInto++;
+        return true;
+    }
+
+    // Computes the greatest common divisor of the absolute values of the operands.
+    // This gets called only if there is at least one negative dynamic offset.
+    // Positive values assume non-negative multipliers and thus enable the precision of the previous analysis.
+    // Negative values indicate that the multiplier can also be negative.
+    private static int reduceAbsGCD(List<Integer> alignment) {
+        if (alignment.isEmpty()) {
+            return 0;
+        }
+        int result = Math.abs(alignment.get(0));
+        for (final Integer a : alignment.subList(1, alignment.size())) {
+            result = IntMath.gcd(result, Math.abs(a));
+        }
+        return result;
+    }
+
+    // Computes the greatest common divisor of the operands.
+    private static int reduceGCD(List<Integer> alignment) {
+        if (alignment.isEmpty()) {
+            return 0;
+        }
+        int result = alignment.get(0);
+        for (final Integer a : alignment.subList(1, alignment.size())) {
+            result = IntMath.gcd(result, a);
+        }
+        return result;
+    }
+
+    // Merges two sets of pairwise-indivisible dynamic offsets.
+    private static List<Integer> compose(List<Integer> left, List<Integer> right) {
+        if (left.isEmpty() || right.isEmpty()) {
+            return right.isEmpty() ? left : right;
+        }
+        if (left == TOP || right == TOP) {
+            return TOP;
+        }
+        // assert left and right each consist of pairwise indivisible positives
+        final List<Integer> result = new ArrayList<>();
+        for (final Integer i : left) {
+            if (!right.contains(i) && hasNoDivisorsInList(i, right)) {
+                result.add(i);
+            }
+        }
```

Comment:
This is a bug.  It was supposed to be a guard to avoid adding common values twice.  `hasNoDivisorsInList` was supposed to return `true` if the list contained the value.